Add fofb orbit setpoints

guilhermerc opened the pull request.

guilhermerc committed:
Updates according to lnls-dig/fofb-ctrl-gw@adac741

guilhermerc committed:
Support fofb orbit setpoints ram

guilhermerc committed:
Some improvements

examples/src/fofb_processing.c:
* add filename argument to '--get-{coeffs,setpoints}' ('-{g,r}');
* remove duplicated 'fixed_point_coeffs' instantiation;
* remove unnecessary 'int32_t_setpoints' instantiation;
* only reads coefficients fixed-point position from hw when necessary

src/client/src/halcs_client_core.c:
* nullified unused 'output' pointer;
* wrapped fofb processing related lines at 80 chars

src/sm_io_table/modules/fofb_processing/sm_io_fofb_processing_exports.c:
* remove unused return value from
  'fofb_processing_coeffs_ram_bank_write_exp'

src/sm_io/modules/fofb_processing/sm_io_fofb_processing_exp.c:
* checked 'ret' pointer in
  'halcs_fofb_processing_coeffs_ram_bank_{read,write}'

ericonr commented on examples/src/fofb_processing.c:
```diff
@@ -212,6 +227,66 @@ int main(int argc, char *argv[]) {
+        break;
+
+      case 'w':
+        errno = 0;
```
nit, why set `errno=0`?

guilhermerc replied:
Indeed, unnecessary...

ericonr commented on examples/src/fofb_processing.c:
```diff
@@ -337,11 +420,61 @@ int main(int argc, char *argv[]) {
+      } else {
+        // casting uint32_t to int32_t
+        for(int i = 0; i < FOFB_PROCESSING_DATA_BLOCK_MAX_PARAMS; i++) {
+          if(fprintf(fp, "%d\n", (int32_t)setpoints.data[i]) < 0) {
```
Nit, you want to use `(int)(int32_t)setpoints.data[i]`; this is unlikely to be an issue, but it's technically more correct.

ericonr commented on include/hw/wb_fofb_processing_regs.h:
```diff
@@ -23,17 +23,23 @@
+#define FOFB_PROCESSING_REGS_COEFFS_RAM_BANK_OFFS(chan)                               \
+  (WB_FOFB_PROCESSING_REGS_COEFFS_RAM_BANK_0 + FOFB_PROCESSING_REGS_COEFFS_RAM_BANK_SIZE*chan)
+
+#define FOFB_PROCESSING_REGS_SETPOINTS_RAM_BANK_SIZE                           \
+  (WB_FOFB_PROCESSING_REGS_SETPOINTS_RAM_BANK - WB_FOFB_PROCESSING_REGS_COEFFS_RAM_BANK_11)
```
Isn't this calculating the size of the 11th coeff bank instead?

guilhermerc replied:
Yeap, it is, both have the same size. Actually, all ram banks have the same size. I'll change it to
`#define FOFB_PROCESSING_REGS_SETPOINTS_RAM_BANK_SIZE FOFB_PROCESSING_REGS_COEFFS_RAM_BANK_SIZE`

ericonr commented on include/hw/wb_fofb_processing_regs.h:
```diff
@@ -23,17 +23,23 @@
+#define FOFB_PROCESSING_REGS_COEFFS_FIXED_POINT_POS_MASK    WB_FOFB_PROCESSING_REGS_COEFFS_FIXED_POINT_POS_VAL_MASK
+#define FOFB_PROCESSING_REGS_COEFFS_FIXED_POINT_POS_SHIFT   WB_FOFB_PROCESSING_REGS_COEFFS_FIXED_POINT_POS_VAL_SHIFT
+#define FOFB_PROCESSING_REGS_COEFFS_FIXED_POINT_POS_R(val)  ((val & FOFB_PROCESSING_REGS_COEFFS_FIXED_POINT_POS_MASK) >> FOFB_PROCESSING_REGS_COEFFS_FIXED_POINT_POS_SHIFT)
+#define FOFB_PROCESSING_REGS_COEFFS_FIXED_POINT_POS_W(val)  ((val & FOFB_PROCESSING_REGS_COEFFS_FIXED_POINT_POS_MASK) << FOFB_PROCESSING_REGS_COEFFS_FIXED_POINT_POS_SHIFT)
```
We shouldn't need to use this macro, since this register is read only, but the definition is wrong; it should be `(val << shift) & mask`.

guilhermerc committed:
Some small changes/fixes

examples/src/fofb_processing.c:
* clear unnecessary 'errno' zeroing;
* cast setpoints to 'int', which is the type expected by '%d' format
  specifier.

include/hw/wb_fofb_processing_regs.h:
* change FOFB_PROCESSING_REGS_SETPOINTS_RAM_BANK_SIZE to a less
  confusing definition;
* fix FOFB_PROCESSING_REGS_COEFFS_FIXED_POINT_POS_W definition.
